clang/lib/Lex/PPDirectives.cpp

```diff
@@ -177,6 +177,26 @@ static bool isLanguageDefinedBuiltin(const SourceManager &SourceMgr,
   return false;
 }
 
+static bool isReservedAttrName(Preprocessor &PP, IdentifierInfo *II) {
```

Comment:
Can you reference some wording here?
Are those hardcoded names sufficient?

Reply:
From my, attribute names aren't typical language keywords, and only a specific set defined by the standard should be disallowed. For example, `likely` and `unlikely` should be permitted. Reusing attribute parsing doesn't seem appropriate here. I haven't found a predefined list of standard attributes; if there is one, I would appreciate it if you could direct me to it. Thanks

Reply:
`likely` is reserved, `likely()` is not.
Also, as we support attributes in older language modes as extensions so we should take that into account.
Does C has similar restrictions?

Reply:
> Can we tablegen that somehow?

@cor3ntin Oke, I'll check...

Reply:
@cor3ntin  The [Attr.td](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/Attr.td) file contain some details about attributes. I'm uncertain whether this file is sufficient for our case or if we need to define a new one...

Reply:
> Does C has similar restrictions?

No, C has no such restrictions.

Reply:
> The [Attr.td](https://github.com/llvm/llvm-project/blob/main/clang/include/clang/Basic/Attr.td) file contain some details about attributes. I'm uncertain whether this file is sufficient for our case or if we need to define a new one...

From what I can tell, you’ll have to extend ClangAttrEmitter and introduce a new flag in `Attr.td` that is set for every standard C++ attribute (e.g. `CXX11NoReturn`, but not `C11NoReturn` or `NoReturn`).

Reply:
@Sirraide @AaronBallman In the context of avoiding hardcoded names., I noticed there's an existing emitter that allows getting `AttrKind` using a string name. Is it okay to use this in this case, or would it be better to add a new emitter that will emit attributes with the new _flag-controlling_ names from the C++ standard?

https://github.com/llvm/llvm-project/blob/9edd998e10fabfff067b9e6e5b044f85a24d0dd5/clang/utils/TableGen/ClangAttrEmitter.cpp#L4875

currenly, it's related to `Sema`, but it might need to be moved to `Basic`...

https://github.com/llvm/llvm-project/blob/9edd998e10fabfff067b9e6e5b044f85a24d0dd5/clang/include/clang/Sema/CMakeLists.txt#L11-L14

Reply:
@Sirraide @AaronBallman @cor3ntin In the context of avoiding hardcoded names., I noticed there's an existing emitter that allows getting `AttrKind` using a string name. Is it okay to use this in this case, or would it be better to add a new emitter that will emit attributes with the new _flag-controlling_ names from the C++ standard?

https://github.com/llvm/llvm-project/blob/9edd998e10fabfff067b9e6e5b044f85a24d0dd5/clang/utils/TableGen/ClangAttrEmitter.cpp#L4875

currenly, it's related to `Sema`, but it might need to be moved to `Basic`...

https://github.com/llvm/llvm-project/blob/9edd998e10fabfff067b9e6e5b044f85a24d0dd5/clang/include/clang/Sema/CMakeLists.txt#L11-L14

Reply:
@AaronBallman Thanks. Just to confirm, are future attribute names valid in previous versions? For instance, can `C++11` use the `deprecated` attribute from `C++14`?

Reply:
Yes, that's a supported extension.

Reply:
The reason I asked is that I recall clang has a predefined macro, `__has_cpp_attribute`, which relies on `hasAttribute`. This function uses a list of existing attributes generated from `Attr.td` for different _`Syntax`_. Should attributes currently defined in clang (for example, `indeterminate` isn’t defined yet) be disallowed, or should all cpp attribute names from the standard be considered?

--- 

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Lex/PPMacroExpansion.cpp#L1915-L1920

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Basic/Attributes.cpp#L31

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Basic/Attributes.cpp#L26

```
...
if (ScopeName == "") {
  return llvm::StringSwitch<int>(Name)
    .Case("noreturn", 200809)
    .Case("assume", 202207)
    .Case("carries_dependency", 200809)
    .Case("deprecated", 201309)
    .Case("fallthrough", 201603)
    .Case("likely", 201803)
    .Case("no_unique_address", true && ( Target.getCXXABI().isItaniumFamily() ) ? 201803 : 0)
    .Case("unlikely", 201803)
    .Case("maybe_unused", 201603)
    .Case("nodiscard", 201907)
    .Default(0);
...
```

Reply:
The reason I asked is that I recall clang has a predefined macro, `__has_cpp_attribute`, which relies on `hasAttribute`. This function uses a list of existing attributes generated from `Attr.td` for different _`Syntax`_. Should attributes currently defined in clang (for example, `indeterminate` isn’t defined yet) be disallowed _(maybe `hasAttribute` could be reused)_, or should all cpp attribute names from the standard be considered _(even if Clang doesn’t fully support it)_?

--- 

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Lex/PPMacroExpansion.cpp#L1915-L1920

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Basic/Attributes.cpp#L31

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Basic/Attributes.cpp#L26

```
...
if (ScopeName == "") {
  return llvm::StringSwitch<int>(Name)
    .Case("noreturn", 200809)
    .Case("assume", 202207)
    .Case("carries_dependency", 200809)
    .Case("deprecated", 201309)
    .Case("fallthrough", 201603)
    .Case("likely", 201803)
    .Case("no_unique_address", true && ( Target.getCXXABI().isItaniumFamily() ) ? 201803 : 0)
    .Case("unlikely", 201803)
    .Case("maybe_unused", 201603)
    .Case("nodiscard", 201907)
    .Default(0);
...
```

--- 

Should this one handle the standard versions too?
https://github.com/llvm/llvm-project/blob/837ee5b46a5f7f898f0de7e46a19600b896a0a1f/clang/lib/Parse/ParseDeclCXX.cpp#L4689-L4708

Reply:
@philnik777 @AaronBallman The reason I asked is that I recall clang has a predefined macro, `__has_cpp_attribute`, which relies on `hasAttribute`. This function uses a list of existing attributes generated from `Attr.td` for different _`Syntax`_. Should attributes currently defined in clang (for example, `indeterminate` isn’t defined yet) be disallowed _(maybe `hasAttribute` could be reused)_, or should all cpp attribute names from the standard be considered _(even if Clang doesn’t fully support it)_?

--- 

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Lex/PPMacroExpansion.cpp#L1915-L1920

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Basic/Attributes.cpp#L31

https://github.com/llvm/llvm-project/blob/357bd61744bb8cc2b9b07447294fa977e5758550/clang/lib/Basic/Attributes.cpp#L26

```
...
if (ScopeName == "") {
  return llvm::StringSwitch<int>(Name)
    .Case("noreturn", 200809)
    .Case("assume", 202207)
    .Case("carries_dependency", 200809)
    .Case("deprecated", 201309)
    .Case("fallthrough", 201603)
    .Case("likely", 201803)
    .Case("no_unique_address", true && ( Target.getCXXABI().isItaniumFamily() ) ? 201803 : 0)
    .Case("unlikely", 201803)
    .Case("maybe_unused", 201603)
    .Case("nodiscard", 201907)
    .Default(0);
...
```

--- 

Should this one handle the standard versions too?
https://github.com/llvm/llvm-project/blob/837ee5b46a5f7f898f0de7e46a19600b896a0a1f/clang/lib/Parse/ParseDeclCXX.cpp#L4689-L4708

Sorry for all the questions., I just want to ensure the flow is clear and duplicates won't be added...

Reply:
I concur with Aaron.  The point of this restriction is to permit the library to use attributes without the risk of someone macroing them away.  

I ALSO think that macroing away non-standards attributes should be perfectly fine, so we should only have this diagnostic for standards attributes.  I THINK we have some infrastructure already to detect those (`__has_cpp_attribute` I think also returns non-zero for non standards attributes?  But perhaps we can re-use the fact that the standards ones use a date?)?

Reply:
> Because we don't support any non-standard attributes without a vendor namespace (e.g., we don't support `[[foo]]` as an extension, it needs to have some sort of prefix like `[[gnu::foo]]` or `[[clang::foo]]`), I was thinking we could rely on that.
> 
> However, it might also make sense to just add a bit to the `Spelling` in `Attr.td` to say "this is a standard spelling for the attribute" (which side-steps things like `nodiscard` vs `warn_unused_result`)?

Ah, yeah, we could just check if it has a non-namespaced standard spelling, thats probably more than enough.  An additional bit isn't really necessary at that point (since we effectively already have it).

Reply:
Maybe? I think they absolutely should, but those names are not portable to other compilers.

(FWIW, I like what the C committee did:

6.7.13.1p4:
> In all aspects of the language, a standard attribute specified by this document as an identifier `attr`
and an identifier of the form `__attr__` shall behave the same when used as an attribute token,
except for the spelling.

so standard library vendors can reliably use the underscored names under an existing well-known identifier reservation rule without sacrificing portability.)

Reply:
Would it make sense to split this into (1) a diagnostic about a name being reserved and (2) a diagnostic about a name being reserved in a future standard?

Reply:
> I think they absolutely should, but those names are not portable to other compilers.

They _should_ be. GCC has always accepted `__attribute__((__foo__))` as well as the `foo` spelling, and uses the same rule for `[[...]]` attributes. I assume Intel compilers do the same for compatibility.

Maybe MSVC doesn't? Seems like something they might support if asked to though.

Reply:
They don't support it: https://godbolt.org/z/9dxshKxj9

> Especially if that rule has been standardized by C.

Yeah, I should probably write a paper for WG21 proposing this.

Reply:
Hmm.... could any such paper remove the above text from `Macro.names`?  It would be nice to be able to abandon this patch as a result.  I could see if this is something we could get through EWG as a DR.

Reply:
That's my intent; I'll send around the paper to you and @jwakely for opinions (I'm most of the way done drafting it). If we think this idea has legs in WG21, then I agree we should hold off on landing this patch as it may be unnecessary.